File: mts/src/mts/execution/improvement_loop.py
"""Multi-step improvement loop for agent tasks.

Orchestrates: generate -> judge -> revise -> judge -> ... -> done.
Stops when quality_threshold is met or max_rounds is exhausted.
"""

from __future__ import annotations

import logging
from dataclasses import dataclass, field

from mts.scenarios.agent_task import AgentTaskInterface

logger = logging.getLogger(__name__)


@dataclass(slots=True)
class RoundResult:
    """Result from a single improvement round."""

    round_number: int
    output: str
    score: float
    reasoning: str
    dimension_scores: dict[str, float] = field(default_factory=dict)
    is_revision: bool = False


@dataclass(slots=True)
class ImprovementResult:
    """Result from the full improvement loop."""

    rounds: list[RoundResult]
    best_output: str
    best_score: float
    best_round: int
    total_rounds: int
    met_threshold: bool

    @property
    def improved(self) -> bool:
        """Whether the final score is higher than the initial score."""
        if len(self.rounds) < 2:
            return False
        return self.rounds[-1].score > self.rounds[0].score


class ImprovementLoop:
    """Orchestrates multi-round improvement of agent task outputs.

    Each round:
    1. Evaluate current output with the judge
    2. If score >= threshold or max rounds reached, stop
    3. Call task.revise_output() with judge feedback
    4. Repeat with revised output
    """

    def __init__(
        self,
        task: AgentTaskInterface,
        max_rounds: int = 5,
        quality_threshold: float = 0.9,
    ) -> None:
        self.task = task
        self.max_rounds = max(1, max_rounds)
        self.quality_threshold = quality_threshold

    def run(
        self,
        initial_output: str,
        state: dict,
        reference_context: str | None = None,
        required_concepts: list[str] | None = None,
        calibration_examples: list[dict] | None = None,
    ) -> ImprovementResult:
        """Run the improvement loop."""
        rounds: list[RoundResult] = []
        current_output = initial_output
        best_output = initial_output
        best_score = 0.0
        best_round = 1

        for round_num in range(1, self.max_rounds + 1):
            logger.info("improvement loop round %d/%d", round_num, self.max_rounds)

            result = self.task.evaluate_output(
                current_output,
                state,
                reference_context=reference_context,
                required_concepts=required_concepts,
                calibration_examples=calibration_examples,
            )

            round_result = RoundResult(
                round_number=round_num,
                output=current_output,
                score=result.score,
                reasoning=result.reasoning,
                dimension_scores=result.dimension_scores,
                is_revision=round_num > 1,
            )
            rounds.append(round_result)

            if result.score > best_score:
                best_score = result.score
                best_output = current_output
                best_round = round_num

            logger.info(
                "round %d score: %.2f (best: %.2f at round %d)",
                round_num, result.score, best_score, best_round,
            )

            if result.score >= self.quality_threshold:
                logger.info("quality threshold %.2f met at round %d", self.quality_threshold, round_num)
                return ImprovementResult(
                    rounds=rounds,
                    best_output=best_output,
                    best_score=best_score,
                    best_round=best_round,
                    total_rounds=round_num,
                    met_threshold=True,
                )

            if round_num < self.max_rounds:
                revised = self.task.revise_output(current_output, result, state)
                if revised == current_output:
                    logger.info("revise_output returned unchanged output, stopping")
                    break
                current_output = revised

        return ImprovementResult(
            rounds=rounds,
            best_output=best_output,
            best_score=best_score,
            best_round=best_round,
            total_rounds=len(rounds),
            met_threshold=False,
        )

File: ts/src/execution/improvement-loop.ts
/**
 * Multi-step improvement loop for agent tasks.
 * Port of mts/src/mts/execution/improvement_loop.py
 */

import type {
  AgentTaskInterface,
  AgentTaskResult,
  RoundResult,
  ImprovementResult,
} from "../types/index.js";

const PARSE_FAILURE_MARKERS = [
  "no parseable score found",
  "missing JUDGE_RESULT markers",
  "invalid JSON",
  "Failed to parse judge response",
] as const;

export function isParseFailure(score: number, reasoning: string): boolean {
  if (score > 0) return false;
  return PARSE_FAILURE_MARKERS.some((m) => reasoning.includes(m));
}

export function isImproved(rounds: RoundResult[]): boolean {
  const valid = rounds.filter((r) => !r.judgeFailed);
  if (valid.length < 2) return false;
  return valid[valid.length - 1].score > valid[0].score;
}

export interface ImprovementLoopOpts {
  task: AgentTaskInterface;
  maxRounds?: number;
  qualityThreshold?: number;
}

export class ImprovementLoop {
  private task: AgentTaskInterface;
  private maxRounds: number;
  private qualityThreshold: number;

  constructor(opts: ImprovementLoopOpts) {
    this.task = opts.task;
    this.maxRounds = Math.max(1, opts.maxRounds ?? 5);
    this.qualityThreshold = opts.qualityThreshold ?? 0.9;
  }

  async run(opts: {
    initialOutput: string;
    state: Record<string, unknown>;
    referenceContext?: string;
    requiredConcepts?: string[];
    calibrationExamples?: Array<Record<string, unknown>>;
  }): Promise<ImprovementResult> {
    const rounds: RoundResult[] = [];
    let currentOutput = opts.initialOutput;
    let bestOutput = opts.initialOutput;
    let bestScore = 0;
    let bestRound = 1;
    let judgeFailures = 0;
    let lastGoodResult: RoundResult | null = null;
    let consecutiveFailures = 0;
    const maxConsecutiveFailures = 3;

    for (let roundNum = 1; roundNum <= this.maxRounds; roundNum++) {
      const result = await this.task.evaluateOutput(currentOutput, opts.state, {
        referenceContext: opts.referenceContext,
        requiredConcepts: opts.requiredConcepts,
        calibrationExamples: opts.calibrationExamples,
      });

      const failed = isParseFailure(result.score, result.reasoning);

      const roundResult: RoundResult = {
        roundNumber: roundNum,
        output: currentOutput,
        score: result.score,
        reasoning: result.reasoning,
        dimensionScores: result.dimensionScores,
        isRevision: roundNum > 1,
        judgeFailed: failed,
      };
      rounds.push(roundResult);

      if (failed) {
        judgeFailures++;
        consecutiveFailures++;

        if (consecutiveFailures >= maxConsecutiveFailures) break;

        if (roundNum < this.maxRounds) {
          if (lastGoodResult && this.task.reviseOutput) {
            const feedbackResult: AgentTaskResult = {
              score: lastGoodResult.score,
              reasoning: lastGoodResult.reasoning,
              dimensionScores: lastGoodResult.dimensionScores,
            };
            const revised = await this.task.reviseOutput(
              currentOutput,
              feedbackResult,
              opts.state,
            );
            if (revised !== currentOutput) currentOutput = revised;
          }
          // else: no prior feedback, just re-judge next round
        }
        continue;
      }

      // Successful evaluation
      consecutiveFailures = 0;
      lastGoodResult = roundResult;

      if (result.score > bestScore) {
        bestScore = result.score;
        bestOutput = currentOutput;
        bestRound = roundNum;
      }

      if (result.score >= this.qualityThreshold) {
        return {
          rounds,
          bestOutput,
          bestScore,
          bestRound,
          totalRounds: roundNum,
          metThreshold: true,
          judgeFailures,
        };
      }

      if (roundNum < this.maxRounds && this.task.reviseOutput) {
        const revised = await this.task.reviseOutput(
          currentOutput,
          result,
          opts.state,
        );
        if (revised === currentOutput) break;
        currentOutput = revised;
      }
    }

    return {
      rounds,
      bestOutput,
      bestScore,
      bestRound,
      totalRounds: rounds.length,
      metThreshold: false,
      judgeFailures,
    };
  }
}
